### DISCUSSION TITLE
Apply transforms to label that has more than one channel

### DISCUSSION THREAD

**Discussion by Alberto1404** [Project-MONAI/MONAI#4225] [Q&A]
Hello everyone. I am still a beginer at MONAI, and I am facing a problem I don't know how to solve. 
To define the train and validation transforms for creating the data loaders, I am doing as follows: 
`from monai.transforms import (
	EnsureChannelFirstd,
	Compose,
	LoadImaged,
	Orientationd,
	RandFlipd,
	RandRotated,
	RandZoomd,
	NormalizeIntensityd,
	ToTensord
)`
The issue comes when dealing with my label. Unlike other common situations, where the shape is (H,W,D) and each voxel takes one scalar, my label has a shape of (3,H,W,D), where each voxel is a array of three values. When applying the transformations (either training or validation), I am getting this warning: 
`Loading dataset:   0%|                                                                                                                        | 0/23 [00:00<?, ?it/s]/home/guijosa/Documents/PythonDocs/PythonDocs_env/lib/python3.8/site-packages/monai/data/utils.py:536: UserWarning: Modifying image pixdim from [1. 1. 1. 0.] to [  1.           1.           1.         363.37824646]
  warnings.warn(f"Modifying image pixdim from {pixdim} to {norm}")
/home/guijosa/Documents/PythonDocs/PythonDocs_env/lib/python3.8/site-packages/monai/data/utils.py:536: UserWarning: Modifying image pixdim from [1. 1. 1. 0.] to [  1.           1.           1.         364.24922786]
  warnings.warn(f"Modifying image pixdim from {pixdim} to {norm}")
/home/guijosa/Documents/PythonDocs/PythonDocs_env/lib/python3.8/site-packages/monai/data/utils.py:536: UserWarning: Modifying image pixdim from [1. 1. 1. 0.] to [  1.           1.           1.         363.54023161]
  warnings.warn(f"Modifying image pixdim from {pixdim} to {norm}")
/home/guijosa/Documents/PythonDocs/PythonDocs_env/lib/python3.8/site-packages/monai/data/utils.py:536: UserWarning: Modifying image pixdim from [1. 1. 1. 0.] to [  1.           1.           1.         364.37686535]
  warnings.warn(f"Modifying image pixdim from {pixdim} to {norm}")
Loading dataset:   4%|████▊                                                                                                           | 1/23 [00:02<00:53,  2.44s/it]/home/guijosa/Documents/PythonDocs/PythonDocs_env/lib/python3.8/site-packages/monai/data/utils.py:536: UserWarning: Modifying image pixdim from [1. 1. 1. 0.] to [  1.           1.           1.         363.94195966]
  warnings.warn(f"Modifying image pixdim from {pixdim} to {norm}")
/home/guijosa/Documents/PythonDocs/PythonDocs_env/lib/python3.8/site-packages/monai/data/utils.py:536: UserWarning: Modifying image pixdim from [1. 1. 1. 0.] to [  1.           1.           1.         366.44576679]
  warnings.warn(f"Modifying image pixdim from {pixdim} to {norm}")
/home/guijosa/Documents/PythonDocs/PythonDocs_env/lib/python3.8/site-packages/monai/data/utils.py:536: UserWarning: Modifying image pixdim from [1. 1. 1. 0.] to [  1.           1.           1.         365.27215881]
  warnings.warn(f"Modifying image pixdim from {pixdim} to {norm}")
/home/guijosa/Documents/PythonDocs/PythonDocs_env/lib/python3.8/site-packages/monai/data/utils.py:536: UserWarning: Modifying image pixdim from [1. 1. 1. 0.] to [  1.         1.         1.       363.651619]
  warnings.warn(f"Modifying image pixdim from {pixdim} to {norm}")
Loading dataset:  35%|██████████████████████████████████████▉                                                                         | 8/23 [00:04<00:06,  2.37it/s]/home/guijosa/Documents/PythonDocs/PythonDocs_env/lib/python3.8/site-packages/monai/data/utils.py:536: UserWarning: Modifying image pixdim from [1. 1. 1. 0.] to [  1.           1.           1.         364.00206043]
  warnings.warn(f"Modifying image pixdim from {pixdim} to {norm}")
/home/guijosa/Documents/PythonDocs/PythonDocs_env/lib/python3.8/site-packages/monai/data/utils.py:536: UserWarning: Modifying image pixdim from [1. 1. 1. 0.] to [  1.           1.           1.         364.50720157]
  warnings.warn(f"Modifying image pixdim from {pixdim} to {norm}")
Loading dataset:  39%|███████████████████████████████████████████▊                                                                    | 9/23 [00:06<00:09,  1.55it/s]/home/guijosa/Documents/PythonDocs/PythonDocs_env/lib/python3.8/site-packages/monai/data/utils.py:536: UserWarning: Modifying image pixdim from [1. 1. 1. 0.] to [  1.           1.           1.         364.06283798]
  warnings.warn(f"Modifying image pixdim from {pixdim} to {norm}")
Loading dataset:  48%|█████████████████████████████████████████████████████                                                          | 11/23 [00:06<00:05,  2.18it/s]/home/guijosa/Documents/PythonDocs/PythonDocs_env/lib/python3.8/site-packages/monai/data/utils.py:536: UserWarning: Modifying image pixdim from [1. 1. 1. 0.] to [  1.           1.           1.         364.77595864]
  warnings.warn(f"Modifying image pixdim from {pixdim} to {norm}")
/home/guijosa/Documents/PythonDocs/PythonDocs_env/lib/python3.8/site-packages/monai/data/utils.py:536: UserWarning: Modifying image pixdim from [1. 1. 1. 0.] to [  1.           1.           1.         365.05547524]
  warnings.warn(f"Modifying image pixdim from {pixdim} to {norm}")
Loading dataset:  57%|██████████████████████████████████████████████████████████████▋                                                | 13/23 [00:07<00:05,  2.00it/s]/home/guijosa/Documents/PythonDocs/PythonDocs_env/lib/python3.8/site-packages/monai/data/utils.py:536: UserWarning: Modifying image pixdim from [1. 1. 1. 0.] to [  1.           1.           1.         380.58080614]
  warnings.warn(f"Modifying image pixdim from {pixdim} to {norm}")
/home/guijosa/Documents/PythonDocs/PythonDocs_env/lib/python3.8/site-packages/monai/data/utils.py:536: UserWarning: Modifying image pixdim from [1. 1. 1. 0.] to [  1.           1.           1.         380.89598318]
  warnings.warn(f"Modifying image pixdim from {pixdim} to {norm}")
Loading dataset:  70%|█████████████████████████████████████████████████████████████████████████████▏                                 | 16/23 [00:08<00:03,  2.15it/s]/home/guijosa/Documents/PythonDocs/PythonDocs_env/lib/python3.8/site-packages/monai/data/utils.py:536: UserWarning: Modifying image pixdim from [1. 1. 1. 0.] to [  1.           1.           1.         364.84483003]
  warnings.warn(f"Modifying image pixdim from {pixdim} to {norm}")
/home/guijosa/Documents/PythonDocs/PythonDocs_env/lib/python3.8/site-packages/monai/data/utils.py:536: UserWarning: Modifying image pixdim from [1. 1. 1. 0.] to [  1.           1.           1.         384.86068908]
  warnings.warn(f"Modifying image pixdim from {pixdim} to {norm}")
/home/guijosa/Documents/PythonDocs/PythonDocs_env/lib/python3.8/site-packages/monai/data/utils.py:536: UserWarning: Modifying image pixdim from [1. 1. 1. 0.] to [  1.           1.           1.         379.04056511]
  warnings.warn(f"Modifying image pixdim from {pixdim} to {norm}")
Loading dataset:  74%|██████████████████████████████████████████████████████████████████████████████████                             | 17/23 [00:10<00:04,  1.39it/s]/home/guijosa/Documents/PythonDocs/PythonDocs_env/lib/python3.8/site-packages/monai/data/utils.py:536: UserWarning: Modifying image pixdim from [1. 1. 1. 0.] to [  1.           1.           1.         382.50718947]
  warnings.warn(f"Modifying image pixdim from {pixdim} to {norm}")
Loading dataset:  87%|████████████████████████████████████████████████████████████████████████████████████████████████▌              | 20/23 [00:11<00:01,  2.31it/s]/home/guijosa/Documents/PythonDocs/PythonDocs_env/lib/python3.8/site-packages/monai/data/utils.py:536: UserWarning: Modifying image pixdim from [1. 1. 1. 0.] to [  1.           1.           1.         381.05445805]
  warnings.warn(f"Modifying image pixdim from {pixdim} to {norm}")
/home/guijosa/Documents/PythonDocs/PythonDocs_env/lib/python3.8/site-packages/monai/data/utils.py:536: UserWarning: Modifying image pixdim from [1. 1. 1. 0.] to [  1.           1.           1.         382.83645333]
  warnings.warn(f"Modifying image pixdim from {pixdim} to {norm}")
/home/guijosa/Documents/PythonDocs/PythonDocs_env/lib/python3.8/site-packages/monai/data/utils.py:536: UserWarning: Modifying image pixdim from [1. 1. 1. 0.] to [  1.           1.           1.         380.11248335]
  warnings.warn(f"Modifying image pixdim from {pixdim} to {norm}")
Loading dataset: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████| 23/23 [00:12<00:00,  1.92it/s]
`

The 4D volume I am saving uses the nifti affine and the nifti header from its correspondant input volume. 
Thank you for your time :)

**Comment by rijobro** [comment]
If you do `image, meta = LoadImage()(label)` where `label` is the filename of one of your label images, what do you get for `image.shape`, and what is the contents of `meta`?

**Reply by Alberto1404** [reply]
Thank you @rijobro for your fast reply. 
`Image.shape = (3, 512, 512, 96)`
The content of meta is shown in the figure below.
<img width="1112" alt="Captura de pantalla 2022-05-04 a las 16 52 33" src="https://user-images.githubusercontent.com/91409881/166708208-27ab8c6d-b297-43ca-9fba-ba850b4a8275.png">

**Reply by rijobro** [reply]
Hi, looks fine to me. What seems to be the problem?

**Reply by Alberto1404** [reply]
In my case I am working with a network that has more than one decoder. Reason why working better with dictionary transforms. I am working in a binary scenario (`out_channels = 1`). The first decoder is the typical semantic segmentator (each value in Ground-Truth takes value 0-background or 1-foreground). When applying transforms here, I get a normal pytorch tensor with (batch_size, channels, H,W,D). Nevertheless, for this second decoder, where my ground-truth (the commented label above) has that 4D-shape, as it is not possible to work with 6D pytorch tensors (batch, channels, ?, H,W,D) I though to directly consider those 3 as literal channels (although we are in a binary scenario). When doing this, I got that warning message applying the transforms to this label.
I don't know whether it will affect to the predictions.

**Reply by rijobro** [reply]
If you are doing binary segmentation, why does your ground truth label consist of 3 channels? What does each channel represent?

You mentioned that you have 6D tensors, but I think you have 4 channels: `C,H,W,D` with shape `3,512,512,96`.

Some ideas:
- Maybe your label is RGB, and each channel dimension is identical to the others?
- Maybe your label contains 3 labels? Background, label 1 and label 2?

Feel free to send your code across and I'll have a look.

**Reply by Alberto1404** [reply]
In that label, each channel represents coordinates x,y and z of vectors located at each voxel position.

**Reply by rijobro** [reply]
Ok, so your label is a deformation or displacement field? And your network is trying to infer that deformation? Is this not just non-rigid registration?
